File: backend/src/pixii_api/llm/nvidia_client.py
import json
import re
from typing import Any, TypeVar

import httpx
import structlog
from pydantic import BaseModel
from tenacity import retry, retry_if_exception, stop_after_attempt, wait_exponential

from pixii_api.config import settings

log = structlog.get_logger()
T = TypeVar("T", bound=BaseModel)

_client: httpx.AsyncClient | None = None


def get_client() -> httpx.AsyncClient:
    global _client
    if _client is None:
        _client = httpx.AsyncClient(
            base_url=settings.nvidia_base_url,
            headers={
                "Authorization": f"Bearer {settings.nvidia_api_key}",
                "Content-Type": "application/json",
            },
            timeout=300.0,  # 5 minutes for LLM API calls
        )
    return _client


def _extract_json(text: str) -> str:
    # Strip markdown code fences if present
    text = text.strip()
    match = re.search(r"```(?:json)?\s*([\s\S]*?)```", text)
    if match:
        return match.group(1).strip()
    # Try to find first { ... } block
    start = text.find("{")
    if start != -1:
        return text[start:]
    return text


def _is_retryable(exc: BaseException) -> bool:
    """Only retry on network errors and 5xx; never retry on 4xx (permanent failures)."""
    if isinstance(exc, httpx.HTTPStatusError):
        return exc.response.status_code >= 500
    return True


@retry(
    stop=stop_after_attempt(3),
    wait=wait_exponential(multiplier=1, min=2, max=10),
    retry=retry_if_exception(_is_retryable),
    reraise=True,
)
async def chat(
    messages: list[dict[str, Any]],
    *,
    model: str | None = None,
    max_tokens: int = 16384,
    temperature: float = 0.7,
) -> str:
    model = model or settings.nvidia_default_model
    payload: dict[str, Any] = {
        "model": model,
        "messages": messages,
        "max_tokens": max_tokens,
        "temperature": temperature,
        "top_p": 1.0,
        "frequency_penalty": 0,
        "presence_penalty": 0,
        "stream": False,
    }

    client = get_client()
    response = await client.post("/chat/completions", json=payload)
    log.debug("nvidia_response", status=response.status_code, model=model)
    response.raise_for_status()
    data = response.json()
    return data["choices"][0]["message"]["content"]


@retry(
    stop=stop_after_attempt(2),
    wait=wait_exponential(multiplier=1, min=2, max=8),
    reraise=True,
)
async def chat_json(
    messages: list[dict[str, Any]],
    schema: type[T],
    *,
    model: str | None = None,
) -> T:
    schema_str = json.dumps(schema.model_json_schema(), indent=2)
    system_msg = {
        "role": "system",
        "content": (
            f"/no_think\n"
            f"You must respond with valid JSON that matches this schema exactly. "
            f"No markdown, no explanation, only JSON.\n\nSchema:\n{schema_str}"
        ),
    }
    full_messages = [system_msg, *messages]

    raw = await chat(full_messages, model=model, temperature=0.3, max_tokens=4096)
    cleaned = _extract_json(raw)

    try:
        return schema.model_validate_json(cleaned)
    except Exception as exc:
        log.warning("json_parse_failed", error=str(exc), raw=raw[:500])
        # Retry with error appended
        repair_messages = [
            *full_messages,
            {"role": "assistant", "content": raw},
            {
                "role": "user",
                "content": (
                    f"That response failed validation: {exc}. "
                    "Return only valid JSON matching the schema."
                ),
            },
        ]
        repair_raw = await chat(repair_messages, model=model, temperature=0.1, max_tokens=4096)
        return schema.model_validate_json(_extract_json(repair_raw))

File: lib/server/nvidiaClient.ts
import { z } from "zod";

const NVIDIA_API_KEY = process.env.NVIDIA_API_KEY!;
const NVIDIA_BASE_URL = "https://integrate.api.nvidia.com/v1";
const NVIDIA_MODEL = "nvidia/nemotron-nano-12b-v2-vl";

type Message = { role: string; content: string | unknown[] };

export function extractJson(text: string): string {
  text = text.trim();
  const fenced = text.match(/```(?:json)?\s*([\s\S]*?)```/);
  if (fenced) return fenced[1].trim();
  const start = text.indexOf("{");
  return start !== -1 ? text.slice(start) : text;
}

async function withRetry<T>(fn: () => Promise<T>, maxAttempts = 3): Promise<T> {
  let lastErr: unknown;
  for (let i = 0; i < maxAttempts; i++) {
    try {
      return await fn();
    } catch (err: unknown) {
      lastErr = err;
      const status = (err as { status?: number }).status;
      if (status && status >= 400 && status < 500) throw err; // don't retry 4xx
      if (i < maxAttempts - 1) await new Promise(r => setTimeout(r, 2000 * (i + 1)));
    }
  }
  throw lastErr;
}

export async function chat(
  messages: Message[],
  opts: { temperature?: number; maxTokens?: number } = {}
): Promise<string> {
  return withRetry(async () => {
    const res = await fetch(`${NVIDIA_BASE_URL}/chat/completions`, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${NVIDIA_API_KEY}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        model: NVIDIA_MODEL,
        messages,
        max_tokens: opts.maxTokens ?? 8192,
        temperature: opts.temperature ?? 0.7,
        top_p: 1.0,
        frequency_penalty: 0,
        presence_penalty: 0,
        stream: false,
      }),
    });

    if (!res.ok) {
      const err = new Error(`NVIDIA API ${res.status}: ${await res.text()}`);
      (err as { status?: number }).status = res.status;
      throw err;
    }

    const data = await res.json();
    return data.choices[0].message.content as string;
  });
}

export async function chatJson<T>(
  messages: Message[],
  schema: z.ZodType<T>,
  opts: { temperature?: number } = {}
): Promise<T> {
  // Serialize schema to JSON Schema so the model knows exact field names
  const jsonSchema = JSON.stringify(z.toJSONSchema(schema), null, 2);

  const systemMsg: Message = {
    role: "system",
    content: `/no_think\nYou must respond with valid JSON that matches this schema exactly. No markdown, no explanation, only JSON.\n\nSchema:\n${jsonSchema}`,
  };

  const fullMessages = [systemMsg, ...messages];
  const raw = await chat(fullMessages, { temperature: opts.temperature ?? 0.3, maxTokens: 8192 });
  const cleaned = extractJson(raw);

  try {
    return schema.parse(JSON.parse(cleaned));
  } catch (err) {
    console.warn("[chatJson] parse failed, attempting repair:", String(err).slice(0, 200));
    const repairMessages: Message[] = [
      ...fullMessages,
      { role: "assistant", content: raw },
      {
        role: "user",
        content: `That response failed validation: ${err}. Return only valid JSON matching the schema exactly.`,
      },
    ];
    const repairRaw = await chat(repairMessages, { temperature: 0.1, maxTokens: 8192 });
    return schema.parse(JSON.parse(extractJson(repairRaw)));
  }
}
